gems/smithy/lib/smithy/views/client/protocol_spec.rb

```diff
@@ -0,0 +1,239 @@
+
+        def module_name
+          @plan.module_name
+        end
+
+        def additional_requires
+          Set.new(@all_operation_tests.map(&:additional_requires).flatten)
+        end
+
+        # @api private
+        class OperationTests
+          def initialize(model, id, operation)
+            @model = model
+            @id = id
+            @operation = operation
+            # TODO: Should we filter protocol tests further by default protocol?
+            # All cases currently have only a single protocol
+            @request_tests = build_request_tests
+            @response_tests = build_response_tests
+            @error_tests = build_error_tests
+          end
+
+          attr_reader :request_tests, :response_tests, :error_tests
+
+          def name
+            Model::Shape.name(@id).underscore
+          end
+
+          def additional_requires
+            @request_tests.map(&:additional_requires) + @response_tests.map(&:additional_requires)
+          end
+
+          def empty?
+            request_tests.empty? && response_tests.empty? && error_tests.empty?
+          end
+
+          private
+
+          def build_response_tests
+            @operation
+              .fetch('traits', {})
+              .fetch('smithy.test#httpResponseTests', [])
+              .select { |t| t.fetch('appliesTo', 'client') == 'client' }
+              .map { |t| ResponseTest.new(@model, @operation, t) }
+          end
+
+          def build_request_tests
+            @operation
+              .fetch('traits', {})
+              .fetch('smithy.test#httpRequestTests', [])
+              .select { |t| t.fetch('appliesTo', 'client') == 'client' }
+              .map { |t| RequestTest.new(@model, @operation, t) }
+          end
+
+          def build_error_tests
+            @operation
+              .fetch('errors', [])
+              .map { |e| tests_for_error(e) }
+              .flatten
+          end
+
+          def tests_for_error(error)
+            error_shape = Model.shape(@model, error['target'])
+            # NOTE: error tests do not have a appliesTo that should be checked
+            error_shape
+              .fetch('traits', {})
+              .fetch('smithy.test#httpResponseTests', [])
+              .map { |t| ErrorTest.new(@model, @operation, error['target'], t) }
+          end
+        end
+
+        # @api private
+        class TestCase
+          def initialize(model, operation, test_case)
+            @model = model
+            @operation = operation
+            @test_case = test_case
+            @input_shape = Model.shape(@model, @operation['input']['target'])
+            @output_shape = Model.shape(@model, @operation['output']['target'])
+          end
+
+          attr_reader :test_case
+
+          def [](key)
+            test_case[key]
+          end
+
+          def comments
+            if test_case['documentation']
+              test_case['documentation'].split("\n")
```

Comment:
Fetch()

Reply:
The code got messier (IMO) with fetch:

```
test_case.fetch('documentation', nil)&.split("\n")
```

Reply:
With a default of an empty array? ..

Reply:
We need to apply split on the string not an array

Reply:
Yeah, you can default fetch to an empty string so that split always returns an array.

Reply:
Oh, right.  Updated.